sirepo/auth_role_moderation.py

```diff
@@ -0,0 +1,105 @@
+from pykern import pkjinja
+import sirepo.api_perm
+import sirepo.auth_db
+import sirepo.http_reply
+import sirepo.http_request
+import sirepo.simulation_db
+
+_cfg = None
+
+@sirepo.api_perm.require_user
+def api_saveModerationReason():
+    def _moderation_link(sim_type):
+        return flask.url_for('_dispatch_empty', _external=True).rstrip('/') \
+               + sirepo.uri.local_route(sim_type, 'admRoles')
+
+    def _dns_reverse_lookup(ip):
+        import dns.resolver, dns.reversename
+        try:
+            return str(dns.resolver.query(dns.reversename.from_address(ip), 'PTR')[0])
+        except (dns.resolver.NoAnswer, dns.resolver.NXDOMAIN):
+            return 'No Reverse DNS Lookup'
+
+    def _send_moderation_email(info):
+        sirepo.smtp.send(
+            recipient=_cfg.moderator_email,
+            subject=f'{info.sim_type} Access Request',
+            body=pkjinja.render_resource('moderation_email', info),
+        )
+
+    req = sirepo.http_request.parse_post(require_sim_type=False)
+    d = req.req_data
+    uid = sirepo.auth.logged_in_user()
+
+    sirepo.auth_db.UserRoleInvite(
+        uid=uid,
+        role=sirepo.auth_role.for_sim_type(d.simulationType),
+        status='pending',
+    ).save()
+
+    _send_moderation_email(
+        PKDict(
+            display_name=sirepo.auth_db.UserRegistration.search_by(uid=uid).display_name,
+            email_addr=sirepo.auth.email.AuthEmailUser.search_by(uid=uid).unverified_email,
+            ip_addr=flask.request.remote_addr,
+            domain_name=_dns_reverse_lookup(flask.request.remote_addr),
+            link=_moderation_link(d.simulationType),
+            reason=d.reason,
+            role=sirepo.auth_role.for_sim_type(d.simulationType),
+            sim_type=d.simulationType,
+            uid=uid,
+            user_agent=flask.request.headers.get("User-Agent"),
+        )
+    )
+    return sirepo.http_reply.gen_json_ok()
+
+
+@sirepo.api_perm.require_user
+def api_getPendingUserRoleInvites():
+    sirepo.auth.check_user_is_adm()
+    rows = [row.__dict__ for row in sirepo.auth_db.UserRoleInvite.get_pending_rows()]
+    headers = [key for key in rows[0].keys() if key != '_sa_instance_state'] if rows else []
+    return sirepo.http_reply.gen_json({
+                'rows': rows,
+                'headers': headers,
+            })
+
+
+@sirepo.api_perm.require_user
+def api_approveOrDenyRole():
+    sirepo.auth.check_user_is_adm()
+    def _send_approve_or_deny_email(uid, approved, info):
+        sirepo.smtp.send(
+            recipient=sirepo.auth.get_module('email').AuthEmailUser.search_by(uid=uid).unverified_email,
+            subject=f'{info.role} Access Request' + ('Approved' if approved else 'Denied'),
+            body=pkjinja.render_resource('approved_role_email' if approved else 'denied_role_email', info),
+        )
+    req = sirepo.http_request.parse_post(require_sim_type=False, type=False)
+    uid = req.req_data.uid
```

Comment:
if you save everything in a PKDict, e.g. `p`, and also add `mode='approve'` or `mode='deny'`, then you refer to things like `approved` as `p.mode == 'approve'` and `uid` as `p.uid`. This keeps with the single letter naming, and makes it easier to pass around parameters.

Reply:
84a51a4c4d292b403b922c73411c9bad58de76bf